### PR TITLE
feat: add opentelemetry utility functions and unit tests

### PR DESCRIPTION
**Pull request by zzzming**
It's the second PR to add more utility functions for validate and add metrics under OpenTelemetry.

Unlike Prometheus' Gauge, OpenTelemetry requires a callback function to update Gauge type. This PR provides a set function to take care of the callback function for the user.

OpenTelementry meter requires to be a singleton. This PR introduces a thread safe singleton creation.

Add new unit tests.

Also enable the prometheus for the debugger.

### CODE REVIEW

File: src/backend/base/langflow/services/telemetry/opentelemetry.py [hunk 1]
```diff
@@ -1,30 +1,219 @@
...
+
+class OpenTelemetry(metaclass=ThreadSafeSingletonMeta):
+    """
+    Define any custom metrics here
+    A thread safe singleton class to manage metrics
+    """
+
+    _metrics_registry: Mapping[str, Metric] = dict[str, Metric](
+        {
+            "file_uploads": Metric(
```
Comment (nicoloboschi) [inline src/backend/base/langflow/services/telemetry/opentelemetry.py:97]: I wonder if we can remove the declaration repetition of the name. maybe having a method that creates a Metric object and call it once per metric to register

e.g.
```
_metrics_registry: Mapping[str, Metric] = dict[str, Metric]


def _add_metric(name: str, description: str, type...):
  self._metrics_registry[name] = ...

self._add_metric("file_uploads"....)
```

File: src/backend/base/langflow/services/telemetry/opentelemetry.py [hunk 2]
```diff
@@ -1,30 +1,219 @@
...
+    def observe_histogram(self, metric_name: str, value: float, labels: Mapping[str, str]):
+        if labels is None or len(labels) == 0:
+            raise ValueError("Labels must be provided for Histogram")
+        histogram = self._metrics.get(metric_name)
+        if isinstance(histogram, Histogram):
+            histogram.record(value, labels)
+        else:
+            raise ValueError(f"Metric '{metric_name}' is not a histogram")
+
+    def update_metric(self, metric_name: str, value: float, labels: Mapping[str, str]):
```
[the remark this answers is not in the document]
Comment (jordanrfrazier) [reply src/backend/base/langflow/services/telemetry/opentelemetry.py:206]: In that you'd prefer separate methods like `update_counter, update_guage`? I think that's not a bad idea either.

File: src/backend/base/langflow/services/telemetry/opentelemetry.py [hunk 3]
```diff
@@ -1,30 +1,219 @@
...
+    def __repr__(self):
+        return f"Metric(name='{self.name}', description='{self.description}', type={self.type}, unit='{self.unit}')"
+
+
+class ThreadSafeSingletonMeta(type):
+    """
+    Thread-safe Singleton metaclass
+    """
+
+    _instances = None
```
Comment (jordanrfrazier) [inline src/backend/base/langflow/services/telemetry/opentelemetry.py:77]: I asked chatgpt how singletons are best implemented in python, so feel free to prove it wrong if you know better (I don't). it said using `weakref.WeakValueDictionary()` is preferred, as it allows the garbage collector to remove the singleton instance after deletion.
```
class ThreadSafeSingletonMetaUsingWeakref(type):
    """
    Thread-safe Singleton metaclass using WeakValueDictionary
    """

    _instances = weakref.WeakValueDictionary()
    _lock: threading.Lock = threading.Lock()

    def __call__(cls, *args, **kwargs):
        if cls not in cls._instances:
            with cls._lock:
                if cls not in cls._instances:
                    instance = super(ThreadSafeSingletonMetaUsingWeakref, cls).__call__(*args, **kwargs)
                    cls._instances[cls] = instance
        return cls._instances[cls]
        ```

**Review by jordanrfrazier — APPROVED** [review]

File: src/backend/base/langflow/services/telemetry/opentelemetry.py (hunk 2 above)
[the remark this answers is not in the document]
Comment (zzzming) [reply src/backend/base/langflow/services/telemetry/opentelemetry.py]: agree. removed

File: src/backend/base/langflow/services/telemetry/opentelemetry.py (hunk 3 above)
Comment (zzzming) [reply src/backend/base/langflow/services/telemetry/opentelemetry.py]: this looks better and I switched to this one. I have a test case to cover race condition, although the previous also passed test.

File: src/backend/base/langflow/services/telemetry/opentelemetry.py [hunk 4]
```diff
@@ -1,30 +1,219 @@
...
-            description="The number of requests",
-        )
+        Register a custom metric
+        It is recommended to register all custom metrics in the constructor's _metrics_registry
+        to track all the metrics in one place.
         """
+        raise NotImplementedError("register_metric is not implemented")
+
+    def increment_counter(self, metric_name: str, labels: Mapping[str, str], value: float = 1.0):
+        if labels is None or len(labels) == 0:
```
[the remark this answers is not in the document]
Comment (zzzming) [reply src/backend/base/langflow/services/telemetry/opentelemetry.py]: done

File: src/backend/base/langflow/services/telemetry/opentelemetry.py (hunk 1 above)
Comment (zzzming) [reply src/backend/base/langflow/services/telemetry/opentelemetry.py]: done

**Pull request merged**